ext/curl/interface.c

```diff
@@ -1185,12 +1194,89 @@ PHP_MINIT_FUNCTION(curl)
 		return FAILURE;
 	}
 
+	zend_class_entry ce;
+	INIT_CLASS_ENTRY(ce, "Curl", curl_object_methods);
+	curl_ce = zend_register_internal_class(&ce);
+	curl_ce->ce_flags |= ZEND_ACC_FINAL;
+	curl_ce->create_object = curl_create_object;
+	curl_ce->serialize = zend_class_serialize_deny;
+	curl_ce->unserialize = zend_class_unserialize_deny;
+
+	memcpy(&curl_object_handlers, &std_object_handlers, sizeof(zend_object_handlers));
+	curl_object_handlers.offset = XtOffsetOf(php_curl, std);
+	curl_object_handlers.free_obj = curl_free_obj;
+	curl_object_handlers.get_gc = curl_get_gc;
+	curl_object_handlers.get_constructor = curl_get_constructor;
+	curl_object_handlers.clone_obj = curl_clone_obj;
+
+	curl_multi_register_class();
+	curl_share_register_class();
 	curlfile_register_class();
 
 	return SUCCESS;
 }
 /* }}} */
 
+/* Curl class */
+
+static zend_object *curl_create_object(zend_class_entry *class_type) {
+	php_curl *intern = zend_object_alloc(sizeof(php_curl), class_type);
+
+	zend_object_std_init(&intern->std, class_type);
+	object_properties_init(&intern->std, class_type);
+	intern->std.handlers = &curl_object_handlers;
+
+	return &intern->std;
+}
+
+static zend_function *curl_get_constructor(zend_object *object) {
+	zend_throw_error(NULL, "Cannot directly construct Curl, use curl_init() instead");
+	return NULL;
+}
+
+static zend_object *curl_clone_obj(zend_object *object) {
+	php_curl *ch;
+	CURL *cp;
+	zval *postfields;
+	zend_object *clone_object;
+	php_curl *clone_ch;
+
+	clone_object = curl_create_object(curl_ce);
+	clone_ch = curl_from_obj(clone_object);
+	init_curl_handle(clone_ch);
+
+	ch = curl_from_obj(object);
+	cp = curl_easy_duphandle(ch->cp);
+	if (!cp) {
+		zend_throw_exception(NULL, "Cannot clone unconstructed Curl object", 0);
+		return &clone_ch->std;
+	}
+
+	clone_ch->cp = cp;
+	_php_setup_easy_copy_handlers(clone_ch, ch);
+
+	postfields = &clone_ch->postfields;
+	if (Z_TYPE_P(postfields) != IS_UNDEF) {
+		if (build_mime_structure_from_hash(clone_ch, postfields) != SUCCESS) {
+			zend_throw_exception(NULL, "Cannot clone unconstructed Curl object", 0);
+			return &clone_ch->std;
+		}
+	}
+
+	return &clone_ch->std;
+}
+
+static HashTable *curl_get_gc(zend_object *object, zval **table, int *n)
+{
+	php_curl *curl = curl_from_obj(object);
+
+	zend_get_gc_buffer *gc_buffer = zend_get_gc_buffer_create();
+	zend_get_gc_buffer_add_zval(gc_buffer, &curl->postfields);
```

Comment:
This now needs to also include other objects held by curl. For ext/curl/tests/bug69485.phpt I believe adding func_name to the gc buffer should help.

Reply:
Oh yes, I had this feeling, but I couldn't determine yet if they are really needed.

Reply:
I amended my last commit with the fix. Hopefully, I did the right thing (I added all the `ZVAL`s that I could find). `ext/curl/tests/bug69485.phpt` is fixed indeed, as well as `ext/curl/tests/bug77535.phpt`. Now, only `ext/curl/tests/bug76675.phpt` fails with:

```
001+ 
001- (Received \d+)+
```

Reply:
That *might* hint at a segfault.

Reply:
Unfortunately my curl is too old to run this test. Does `run-tests.php -m ext/curl/tests/bug76675.phpt` give anything useful?

Reply:
Okay, I've built a newer curl now, and the actual output of the test is:

```
Received 3755Received 10830Received 16375Received 9Received 2340Received 1738Received 7683Received 4078Received 12306Received 4078Received 12306Received 16375Received 9Received 16375Received 9Received 16375Received 9Received 11515[Tue Apr 28 11:23:48 2020]  Script:  '/home/nikic/php-src/t180.php'
/home/nikic/php-src/Zend/zend_closures.c(482) :  Freeing 0x00007fcd2be6e900 (312 bytes), script=/home/nikic/php-src/t180.php
Last leak repeated 3 times
[Tue Apr 28 11:23:48 2020]  Script:  '/home/nikic/php-src/t180.php'
/home/nikic/php-src/Zend/zend_objects_API.h(91) :  Freeing 0x00007fcd2bea4000 (368 bytes), script=/home/nikic/php-src/t180.php
Last leak repeated 4 times
=== Total 9 memory leaks detected ===
```

So probably just a GC issue.